In [48]:
#Building Decision Tree Model and Random Forest Algorithm
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import statsmodels.api as sm
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.tree import DecisionTreeClassifier
from sklearn import tree
from IPython.display import Image
from sklearn import preprocessing
from sklearn.ensemble import RandomForestClassifier
import pydotplus
import graphviz 
import warnings

warnings.filterwarnings('ignore')

# Build Decision Tree(DV-"Survived",IDV-"Age,Gender and Fare") and Prediction

In [49]:
dftrain = pd.read_csv('train.csv')
dftest = pd.read_csv('test.csv')
dftrain.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [50]:
dftest.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Embarked  
0  34.5      0      0   330911   7.8292        Q  
1  47.0      1      0   363272   7.0000        S  
2  62.0      0      0   240276   9.6875        Q  
3  27.0      0      0   315154   8.6625        S  
4  22.0      1      1  3101298  12.2875        S

In [51]:
trainage = np.where(dftrain.Age.isnull(), dftrain.Age.mean(), dftrain.Age)
testage = np.where(dftest.Age.isnull(), dftest.Age.mean(), dftest.Age)

In [52]:
label_en = preprocessing.LabelEncoder()

In [53]:
train_gender = label_en.fit_transform(dftrain.Sex)
test_gender = label_en.fit_transform(dftest.Sex)

In [54]:
features = pd.DataFrame([trainage, dftrain.Fare, train_gender]).T

In [55]:
model1 = tree.DecisionTreeClassifier(max_depth = 6)
model1.fit(X = features, y = dftrain.Survived)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=6, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [56]:
model1.score(X = features, y = dftrain.Survived)

0.8301462317210349

0.1  Inference

We get a score of 82.90 % when the features are Age, Fare and Gender



Predication

In [57]:
features_test = pd.DataFrame([testage, dftest.Fare, test_gender]).T

In [58]:
pred = model1.predict(X = features_test)

In [59]:
pred_output = pd.DataFrame({"P_Id": dftest.PassengerId, "Survived" : pred})

In [60]:
pred_output.head()

P_Id  Survived
0   892         0
1   893         1
2   894         0
3   895         0
4   896         1

In [61]:
with open('DecisionTree1.dot','w') as file:
    file = tree.export_graphviz(model1, feature_names=['AGE','FARE','GENDER'], out_file= file)

# 2: Build Decision Tree for Attrition Rate Analysis


DV - "Attrition"

IDV - Output of RF Algorithm


In [62]:
df = pd.read_csv('general_data.csv')

In [63]:
df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [64]:
df['NumCompaniesWorked'] = np.where(df.NumCompaniesWorked.isnull(), df.NumCompaniesWorked.mean(), df.NumCompaniesWorked )
df['TotalWorkingYears'] = np.where(df.TotalWorkingYears.isnull(), df.TotalWorkingYears.mean(), df.TotalWorkingYears )

In [65]:
le = preprocessing.LabelEncoder()

In [66]:
df_temp = df.copy()

In [67]:
df_temp['Attrition'] = le.fit_transform(df.Attrition)
df_temp['BusinessTravel'] = le.fit_transform(df.BusinessTravel)
df_temp['Department'] = le.fit_transform(df.Department)
df_temp['EducationField'] = le.fit_transform(df.EducationField)
df_temp['Gender'] = le.fit_transform(df.Gender)
df_temp['JobRole'] = le.fit_transform(df.JobRole)
df_temp['MaritalStatus'] = le.fit_transform(df.MaritalStatus)
df_temp['Over18'] = le.fit_transform(df.Over18)

In [68]:
model2 = RandomForestClassifier(n_estimators = 2000, max_features = 2, oob_score = True)

In [69]:
attr_train_feats = ['Age', 'BusinessTravel', 'Department', 'DistanceFromHome', 'Education', 'EducationField', 
                'Gender', 'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome', 'NumCompaniesWorked', 
                'Over18', 'PercentSalaryHike', 'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 
                'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager']

In [70]:
model2.fit(X = df_temp[attr_train_feats], y = df_temp.Attrition)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features=2,
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=2000,
                       n_jobs=None, oob_score=True, random_state=None,
                       verbose=0, warm_start=False)

In [71]:
model2.fit(X = df_temp[attr_train_feats], y = df_temp.Attrition)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features=2,
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=2000,
                       n_jobs=None, oob_score=True, random_state=None,
                       verbose=0, warm_start=False)

In [72]:
model2.oob_score_

1.0

In [73]:
for train_feat, imp in zip(attr_train_feats, model2.feature_importances_):
    print(train_feat, '-->', imp)

Age --> 0.09611987200769224
BusinessTravel --> 0.02793770348840805
Department --> 0.025796052207941626
DistanceFromHome --> 0.06997807967915255
Education --> 0.04047582943699926
EducationField --> 0.042073307579852295
Gender --> 0.01808238068705069
JobLevel --> 0.03761714024746734
JobRole --> 0.056104844335579045
MaritalStatus --> 0.03987070197237922
MonthlyIncome --> 0.09344113721727464
NumCompaniesWorked --> 0.05665791075106289
Over18 --> 0.0
PercentSalaryHike --> 0.06562501343568543
StockOptionLevel --> 0.03402889260852531
TotalWorkingYears --> 0.08613961685003796
TrainingTimesLastYear --> 0.04523598877675276
YearsAtCompany --> 0.06802368078901437
YearsSinceLastPromotion --> 0.042584452237029086
YearsWithCurrManager --> 0.05420739569209519


Inference

- Looking at the importance score for each feature we can conclude that features like
> Age, MonthlyIncome, TotalWorkingYears, DistanceFromHome, PercentSalaryHike, YearsAtCompany seems to be most importance among other features.

In [74]:
attr_test_feat = ['Age','DistanceFromHome', 'MonthlyIncome', 'PercentSalaryHike', 'TotalWorkingYears', 'YearsAtCompany']

In [75]:
model3 = tree.DecisionTreeClassifier(max_depth = 12)

In [76]:
model3.fit(X = df_temp[attr_test_feat], y = df_temp.Attrition)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=12, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [77]:
model3.score(X = df_temp[attr_test_feat], y = df_temp.Attrition)

0.9480725623582766

Inference

> With score of 94.80% we can say that Age, MonthlyIncome, TotalWorkingYears, DistanceFromHome, PercentSalaryHike, YearsAtCompany features play important role in attrition level in the company.



# 3 Build Decision Tree for Bank Loan Modelling¶

-    DV - "Personal Loan"
-   IDV - Output of RF Algorithm


In [78]:
df3 = pd.read_excel('Bank_Personal_Loan_Modelling.xlsx', sheet_name='Data')

In [79]:
df3.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [80]:
model4 = RandomForestClassifier(n_estimators = 2000, max_features = 2 , oob_score = True)

In [81]:
bank_train_feats =[ 'Age', 'Experience', 'Income', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Securities Account',
       'CD Account', 'Online', 'CreditCard']

In [82]:
model4.fit(X = df3[bank_train_feats], y = df3['Personal Loan'])

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features=2,
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=2000,
                       n_jobs=None, oob_score=True, random_state=None,
                       verbose=0, warm_start=False)

In [83]:
model4.oob_score_

0.9878

In [84]:
for train_feat, imp in zip(bank_train_feats, model4.feature_importances_):
    print(train_feat, '-->', imp)

Age --> 0.044759838735911446
Experience --> 0.04429287224930486
Income --> 0.34238480819147626
Family --> 0.09535502548556184
CCAvg --> 0.1874476475532184
Education --> 0.16317436879365446
Mortgage --> 0.04445563442653634
Securities Account --> 0.0057309137079656946
CD Account --> 0.05405687423192686
Online --> 0.00849655220170757
CreditCard --> 0.009845464422736344


Inference

>  We can obsevrve that Age, CCAvg, Education, Family, CD Account, Mortage seem to be important features


In [85]:
bank_test_feats = ['Family', 'CCAvg', 'Education', 'CD Account','Age','Mortgage']

In [86]:
model5 = tree.DecisionTreeClassifier(max_depth = 12)

In [87]:
model5.fit(X = df3[bank_test_feats], y = df3['Personal Loan'])

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=12, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [88]:
model5.score(X = df3[bank_test_feats], y = df3['Personal Loan'])

0.9866

Inference

>   With score of 98.66% we can say that Income, CCAvg, Education, Family , CD Account features play important role in deciding whether the person will get a loan or not.

